File: android/src/main/java/com/nitroplay/streaming/HlsRuntimeState.kt
package com.nitroplay.video.streaming

internal data class HlsRuntimeStateSnapshot(
  val port: Int,
  val didAutoStart: Boolean,
  val isExplicitlyStopped: Boolean
)

internal class HlsRuntimeState(
  private val defaultPort: Int = 0
) {
  private val lock = Any()
  private var port: Int = defaultPort
  private var didAutoStart = false
  private var isExplicitlyStopped = false

  fun start(requestedPort: Int?): Int {
    synchronized(lock) {
      val resolvedPort = if ((requestedPort ?: defaultPort) > 0) requestedPort ?: defaultPort else defaultPort
      port = resolvedPort
      isExplicitlyStopped = false
      return resolvedPort
    }
  }

  fun stop() {
    synchronized(lock) {
      didAutoStart = false
      isExplicitlyStopped = true
    }
  }

  fun suspendForHostLifecycle() {
    synchronized(lock) {
      didAutoStart = false
    }
  }

  fun portForImplicitStart(): Int? {
    synchronized(lock) {
      if (isExplicitlyStopped) return null
      if (didAutoStart) return null
      return port
    }
  }

  fun markAutoStarted() {
    synchronized(lock) {
      if (!isExplicitlyStopped) {
        didAutoStart = true
      }
    }
  }

  fun portForPlaybackRecoveryRestart(): Int? {
    synchronized(lock) {
      if (isExplicitlyStopped) return null
      return port
    }
  }

  fun snapshot(): HlsRuntimeStateSnapshot {
    synchronized(lock) {
      return HlsRuntimeStateSnapshot(
        port = port,
        didAutoStart = didAutoStart,
        isExplicitlyStopped = isExplicitlyStopped
      )
    }
  }
}

File: ios/streaming/HlsRuntimeState.swift
import Foundation

struct HlsRuntimeStateSnapshot {
  let port: Int
  let didAutoStart: Bool
  let isExplicitlyStopped: Bool
}

final class HlsRuntimeState {
  private let defaultPort: Int
  private let queue = DispatchQueue(label: "com.nitroplay.hls.runtime-state-machine")

  private var port: Int
  private var didAutoStart = false
  private var isExplicitlyStopped = false

  init(defaultPort: Int = 0) {
    self.defaultPort = defaultPort
    self.port = defaultPort
  }

  func start(port requestedPort: Int?) -> Int {
    queue.sync {
      let resolvedPort = (requestedPort ?? defaultPort) > 0 ? (requestedPort ?? defaultPort) : defaultPort
      port = resolvedPort
      isExplicitlyStopped = false
      return resolvedPort
    }
  }

  func stop() {
    queue.sync {
      isExplicitlyStopped = true
      didAutoStart = false
    }
  }

  func suspendForHostLifecycle() {
    queue.sync {
      didAutoStart = false
    }
  }

  func portForImplicitStart() -> Int? {
    queue.sync {
      guard !isExplicitlyStopped else {
        return nil
      }
      if didAutoStart {
        return nil
      }
      return port
    }
  }

  func markAutoStarted() {
    queue.sync {
      if !isExplicitlyStopped {
        didAutoStart = true
      }
    }
  }

  func portForPlaybackRecoveryRestart() -> Int? {
    queue.sync {
      guard !isExplicitlyStopped else {
        return nil
      }
      return port
    }
  }

  func snapshot() -> HlsRuntimeStateSnapshot {
    queue.sync {
      HlsRuntimeStateSnapshot(
        port: port,
        didAutoStart: didAutoStart,
        isExplicitlyStopped: isExplicitlyStopped
      )
    }
  }
}
